src/rwlock.rs

```diff
@@ -304,161 +501,143 @@ impl<T: fmt::Debug + ?Sized> fmt::Debug for RwLock<T> {
-            lock: self.raw.lock,
-            value: self.value,
-        })
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&**self, f)
     }
 }
 
-/// The future returned by [`RwLock::upgradable_read`].
-pub struct UpgradableRead<'a, T: ?Sized> {
-    /// Raw upgradable read lock acquisition future, doesn't depend on `T`.
-    raw: RawUpgradableRead<'a>,
-
-    /// Pointer to the value protected by the lock. Invariant in `T`
-    /// as the upgradable lock could provide write access.
-    value: *mut T,
-}
-
-unsafe impl<T: Send + Sync + ?Sized> Send for UpgradableRead<'_, T> {}
-unsafe impl<T: Send + Sync + ?Sized> Sync for UpgradableRead<'_, T> {}
-
-impl<T: ?Sized> fmt::Debug for UpgradableRead<'_, T> {
+impl<T: fmt::Display + ?Sized> fmt::Display for RwLockReadGuard<'_, T> {
+    #[inline]
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str("UpgradableRead { .. }")
+        (**self).fmt(f)
     }
 }
 
-impl<T: ?Sized> Unpin for UpgradableRead<'_, T> {}
-
-impl<'a, T: ?Sized> Future for UpgradableRead<'a, T> {
-    type Output = RwLockUpgradableReadGuard<'a, T>;
+impl<T: ?Sized> Deref for RwLockReadGuard<'_, T> {
+    type Target = T;
 
     #[inline]
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        ready!(Pin::new(&mut self.raw).poll(cx));
-
-        Poll::Ready(RwLockUpgradableReadGuard {
-            lock: self.raw.lock,
-            value: self.value,
-        })
+    fn deref(&self) -> &T {
+        unsafe { &*self.value }
     }
 }
 
-/// The future returned by [`RwLock::write`].
-pub struct Write<'a, T: ?Sized> {
-    /// Raw write lock acquisition future, doesn't depend on `T`.
-    raw: RawWrite<'a>,
-
-    /// Pointer to the value protected by the lock. Invariant in `T`.
-    value: *mut T,
+/// An owned, reference-counting guard that releases the read lock when dropped.
+#[clippy::has_significant_drop]
+pub struct ArcRwLockReadGuard<T> {
+    /// **WARNING**: This doesn't actually point to a `T`!
+    /// It points to a `RwLock<T>`, via a pointer obtained with `Arc::into_raw`.
+    /// We lie for covariance.
+    lock: NonNull<T>,
 }
 
-unsafe impl<T: Send + Sync + ?Sized> Send for Write<'_, T> {}
-unsafe impl<T: Send + Sync + ?Sized> Sync for Write<'_, T> {}
+unsafe impl<T: Send + Sync> Send for ArcRwLockReadGuard<T> {}
+unsafe impl<T: Send + Sync> Sync for ArcRwLockReadGuard<T> {}
 
-impl<T: ?Sized> fmt::Debug for Write<'_, T> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str("Write { .. }")
+impl<T> ArcRwLockReadGuard<T> {
+    /// Constructs the underlying `Arc` back from the underlying `RwLock`.
+    ///
+    /// # Safety
+    ///
+    /// Both the returned `Arc` and the guard will decrement their reference
+    /// counts on drop! So one of the two must be forgotten.
+    #[inline]
+    unsafe fn inner_arc(guard: &Self) -> Arc<RwLock<T>> {
```

Comment:
It might be less error-prone to return an `ManuallyDrop<Arc<RwLock<T>>>` instead.